File: Graph/bfs.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

public class bfs {
  public static void main(String[] args) {
    // Dummy graph as an edge list
    int[][] edges = {
        { 0, 1 },
        { 0, 2 },
        { 1, 3 },
        { 1, 4 },
        { 2, 5 },
        { 2, 6 }
    };

    int numNodes = 7;
    ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
    for (int i = 0; i < numNodes; i++) {
      adj.add(new ArrayList<>());
    }

    for (int[] edge : edges) {
      adj.get(edge[0]).add(edge[1]);
      adj.get(edge[1]).add(edge[0]); // undirected graph
    }

    for (int i = 0; i < numNodes; i++) {
      System.out.print(i + " -> ");
      for (int neighbor : adj.get(i)) {
        System.out.print(neighbor + " ");
      }
      System.out.println();
    }

    System.out.println();
    breadthFistSearch(adj);
  }

  public static void breadthFistSearch(ArrayList<ArrayList<Integer>> adj) {
    boolean[] visited = new boolean[adj.size()];
    Queue<Integer> q = new LinkedList<>();
    q.offer(0);
    visited[0] = true;

    while (!q.isEmpty()) {
      int curr = q.poll();
      System.out.println(curr);

      for (int i = 0; i < adj.get(curr).size(); i++) {
        int val = adj.get(curr).get(i);
        if (!visited[val]) {
          visited[val] = true;
          q.offer(val);
        }
      }
    }
  }
}

File: Graph/src/bfs.cpp
#include "../include/graph.hpp"
#include <iostream>
#include <queue>
#include <vector>

std::vector<int> Graph::bfs(std::vector<std::vector<int>> &graph) {
  std::queue<int> queue;
  std::vector<int> bfs;
  std::vector<bool> vis(graph_data::num_nodes, false);
  for (int i = 0; i < graph_data::num_nodes; i++) {
    if (vis[i])
      continue;
    vis[i] = true;
    queue.push(i);
    while (!queue.empty()) {
      int size = queue.size();
      for (int i = 0; i < size; i++) {
        int node = queue.front();
        bfs.emplace_back(node);
        queue.pop();
        for (int neighbor : graph[node]) {
          if (!vis[neighbor]) {
            vis[neighbor] = true;
            queue.push(neighbor);
          }
        }
      }
    }
  }
  return bfs;
}

int main() {
  std::vector<std::vector<int>> graph(graph_data::num_nodes);
  for (auto &edge : graph_data::edges) {
    graph[edge.first].emplace_back(edge.second);
    graph[edge.second].emplace_back(edge.first);
  }
  // int node_id = 0;
  // for (const auto &neighbors : graph) {
  //   std::cout << "Node " << node_id << " -> ";
  //   for (int neighbor : neighbors) {
  //     std::cout << neighbor << " ";
  //   }
  //   std::cout << std::endl;
  //   node_id++;
  // }
  Graph g;
  std::vector<int> result = g.bfs(graph);

  std::cout << "BFS Traversal: ";
  for (int node : result) {
    std::cout << node << " ";
  }
  std::cout << std::endl;

  return 0;
}
